src/mon/PGMonitor.cc

```diff
@@ -1234,6 +1235,146 @@ bool PGMonitor::preprocess_command(MonOpRequestRef op)
                           mon->osdmon()->osdmap.get_inst(osd));
     ss << "instructing pg " << pgid << " on osd." << osd << " to " << scrubop;
     r = 0;
+  } else if (prefix == "pg force-recovery" ||
+             prefix == "pg force-backfill" ||
+             prefix == "pg cancel-force-recovery" ||
+             prefix == "pg cancel-force-backfill") {
+    string forceop = prefix.substr(3, string::npos);
+    list<pg_t> parsed_pgs;
+
+    // figure out actual op just once
+    int actual_op = 0;
+    if (forceop == "force-recovery") {
+      actual_op = OFR_RECOVERY;
+    } else if (forceop == "force-backfill") {
+      actual_op = OFR_BACKFILL;
+    } else if (forceop == "cancel-force-backfill") {
+      actual_op = OFR_BACKFILL | OFR_CANCEL;
+    } else if (forceop == "cancel-force-recovery") {
+      actual_op = OFR_RECOVERY | OFR_CANCEL;
+    }
+
+    // covnert pg names to pgs, discard any invalid ones while at it
+    {
+      // we don't want to keep pgidstr and pgidstr_nodup forever
+      vector<string> pgidstr;
+      set<string> pgidstr_nodup;
+
+      // get pgids to process and prune duplicates
+      cmd_getval(g_ceph_context, cmdmap, "pgid", pgidstr);
+      for (auto& pstr : pgidstr) {
+	pgidstr_nodup.insert(pstr);
+      }
+      if (pgidstr.size() != pgidstr_nodup.size()) {
+	// move elements only when there were duplicates, as this
+	// reorders them
+	pgidstr.resize(pgidstr_nodup.size());
+	auto it = pgidstr_nodup.begin();
+	for (size_t i = 0 ; i < pgidstr_nodup.size(); i++) {
+	  pgidstr[i] = std::move(*it++);
+	}
```

Comment:
this could be simplified to:

    pgidstr.clear();
    pgidstr.insert(pgidstr.end(), make_move_iterator(pgidstr_nodup.begin()), make_move_iterator(pgidstr_nodup.end()));

Technically you should be able to create a new vector from the set (with the elements moved) and then move the new vector into `pgidstr` but not sure how much that would save you. Don't think there is much point in re-sizing the original vector.